### ISSUE TITLE
React Komposer vs createContainer?

### ISSUE THREAD

**Issue by SachaG** [meteor-utilities/react-list-container#3]
I'd like to get rid of `getMeteorData`, but I'm not sure if I should switch to the official `createContainer` or use React Komposer? The syntax is pretty similar so I'm not sure what the differences are.

**Comment by arunoda** [comment]
react-komposer comes with NPM. I will definitely go with anything comes with NPM. 
(Anyway, NPM is a thing these days :D)

I suggest to get rid of Meteor packages as possible as you can.

**Comment by arunoda** [comment]
May be you can make this a a NPM package.

**Comment by SachaG** [comment]
Good point. 1 point for react-komposer then :)

**Comment by yched** [comment]
No strong feeling on my side :-)

On the site I'm building I used a simple version of the forthcoming "official" createContainer HOC, because that's 10 lines of code around getMeteorData, which is fairly stable and doesn't take us away from the "official" ecosystem.

I had a good look at react-komposer back then, but it felt too much for our needs, and didn't feel as stable (again, _back then_)

**Comment by jshimko** [comment]
How about https://github.com/studiointeract/tracker-component ?  The simplicity is quite attractive and I like how similar it is to what we've already been doing with Blaze.  

It's not clear whether there are performance trade-offs though.  Any thoughts on this @arunoda?  (I only ask because you wrote a similar package)  The entire package is essentially [this file](https://github.com/studiointeract/tracker-component/blob/master/lib/index.jsx).

**Comment by SachaG** [comment]
It does look nice, but I have to confess I have no clue about the trade-offs vs React Komposer or createContainer, if there are any. Right now I would tend towards using Komposer just because it seems like the most-used out there, but there's also arguments for using createContainer since it's the default MDG solution.

**Comment by zeroasterisk** [comment]
I love what you're doing here.  

I am very interested in what you end up choosing, because I, too, have had a kinda hard time figuring out [which container approach to choose](https://github.com/studiointeract/tracker-component#comparison).

For now, I'm sticking with the "official" `createContainer`, mostly because of the word "official" but it smells a bit bulky and boilerplate-heavy.  On the other hand, I like @arunoda 's point about `npm` module being a winning factor.

I'm just trying to think about keeping ALL of my meteor concerns/code/data decoupled from the components... Truthfully, whatever containering wrapper you choose should be fine, but I'd like to see an example in the official _guide_ pattern of `imports/ui/(containers,pages,components)/thing.jsx` which used your components.

**Comment by SachaG** [comment]
I tried React Komposer but then got stuck on this issue: https://github.com/kadirahq/react-komposer/issues/69

I haven't looked into it again since, if I find a way to make it work I'll use Komposer, if not I'll use `createContainer`.

**Comment by schabluk** [comment]
I tried to wrap my head around Meteor's createContainer after using komposer, but I'm missing one thing, which is the `onData` callback. I can use it easly with `Meteor.call`.

```
export const composer = ({context}, onData) => {
  const {Meteor} = context()
  Meteor.call('profiles.fetch', {}, function(err, data) {
      onData(null, {data})
  })
}
```

In createContainer there is no callback parameter, so invoking async method will fail to deliver results.

**Comment by zeroasterisk** [comment]
@schabluk 
I could be wrong-as-hell here, and smarter people may correct me here, 
but my understanding is that everything inside a `createContainer` is inside a `Tracker.autorun` and it gets re-run over and over, as anything inside it triggers a recompute (of the container, and the page, and the components, on down... they get new props and React determines if it needs to re-render on the DOM.)

**Comment by schabluk** [comment]
@zeroasterisk 
When it goes to meteor collections, yes, the tracker should trigger refresh and the container will reload, but in my case, the data source is not reactive. 
So the question is, how do I return the result of a `Meteor.call` inside `createContainer`? Consider this example:

```
export default ProfilesContainer = createContainer(({ params }) => {
  Meteor.call('profiles.fetch', {}, (err, result) => {
    // how to return result from callback to the createContainer scope? 
  })
  return {
    data: result // need result from callback here
  }
}, ProfilesComponent)
```

I could resolve this, byt doing a `sync` call, like this:

```
export default ProfilesContainer = createContainer(({ params }) => {
  const result = Meteor.call('profiles.fetch', {})
  return {
    data: result
  }
}, ProfilesComponent)
```

But, according to Meteor docs, that will work only for stubs:
"On the client, if you do not pass a callback and you are not inside a stub, call will return undefined, and you will have no way to get the return value of the method. That is because the client doesn’t have fibers, so there is not actually any way it can block on the remote execution of a method."
http://docs.meteor.com/api/methods.html#Meteor-call

On the server, I can use `Meteor.wrapAsync`, to return a result from async call. Here is my server method:

```
import seraph from 'seraph'

const Neo4j = seraph({
  user: this.process.env.NEO4J_USER,
  pass: this.process.env.NEO4J_PASS
})

Meteor.methods({
  'profiles.fetch'(params) {
    let cypher = "MATCH (p:Profiles) RETURN p"
    let wrapper = Meteor.wrapAsync((params, callback) => {
      Neo4j.query(cypher, params, callback)
    })
    let result = wrapper(params)
    return result
  }
})
```

Unfortunately, this will not work on the client - again the callback is required:
"On the server, the wrapped function can be used either synchronously (without passing a callback) or asynchronously (when a callback is passed). On the client, a callback is always required; errors will be logged if there is no callback."
http://docs.meteor.com/api/core.html#Meteor-wrapAsync

With @arunoda's react-komposer I can solve this on the client side using `onData` function inside the callback:

```
import {useDeps, composeWithTracker, composeAll} from 'mantra-core'

export const composer = ({context}, onData) => {
  const {Meteor} = context()
  Meteor.call('profiles.fetch', {}, function(err, data) {
      onData(null, {data})
  })
}
export default composeAll(
  composeWithTracker(composer),
  useDeps()
)(ProfilesComponent)
```

**Comment by SachaG** [comment]
I haven't read through this whole issue yet, but I just wanted to say I'd be happy to switch this package to React Komposer if somebody can help me figure out https://github.com/kadirahq/react-komposer/issues/69

**Comment by schabluk** [comment]
@SachaG doesn't createContainer use getMeteorData from ReactMeteorData mixin [behind the scene](https://github.com/meteor/react-packages/blob/master/packages/react-meteor-data/createContainer.jsx)?

**Comment by SachaG** [comment]
@schabluk yeah I think so. I guess the difference is more the syntax and pattern used.

**Comment by wiiiimm** [comment]
@schabluk i haven't tried it. but you will using a reactive var to pass data from a callback back out to the createContainer work?

````
export default ProfilesContainer = createContainer(({ params }) => {
  Meteor.call('profiles.fetch', {}, (err, result) => {
    // set reactiveData here
  })
  return {
    data: reactiveData
  }
}, ProfilesComponent)
````

been using `react-meteor-data` for a while now, just came across this issue when researching if it is better to switch over to `react-komposer` XD.
